### ISSUE TITLE
Make a plan for sort_values/set_index

### ISSUE THREAD

**Issue by mrocklin** [rapidsai/dask-cudf#178]
It would be nice to be able to use the `set_index` method to sort the dataframe by a particular column.

There are currently two implementations for this, one in dask.dataframe and one in dask-cudf which uses a batcher sorting net.  While most dask-cudf code has been removed in favor of the dask.dataframe implementations this sorting code has remained, mostly because I don't understand it fully, and don't know if there was a reason for this particular implementation.

Why was this implementation chosen?  Was this discussed somewhere?  Alternatively @sklam, do you have any information here?

cc @kkraus14 @randerzander

**Comment by mrocklin** [comment]
NVIDIA folks have asked if there is some way to integrate an MPI or NCCL enabled multi-gpu sort into Dask for improved efficiency.  My initial reaction to this is that it's likely to be difficult to integrate smoothly in a way that respects other Dask management like resilience, spilling to disk, load balancing, and so on.  Lets expand on this.

First, if we have a multi-node sorting algorithm, we can always treat it how we treat XGBoost.  Dask gives up control to some other higher performance system, it does its thing, we claim control back.  If anything fails during this stage then we just retry the whole thing.  We give up on any kind of memory management or load balancing during this process and just hope that the external system can handle things well without blowing up.

Second question is if we just have a single-node multi-GPU system, maybe we can use that?  This is also a bit tricky currently, but we might be able to make structural changes to Dask to make it less tricky.  The cost-benefit analysis of those changes might make this undesirable though.  Currently the approach most people seem to be using with Dask and GPUs is to have one Dask worker per GPU.  Currently Dask workers don't have any knowledge of other Dask workers on the same node, so there isn't anything built up to handle local collective action.  We would be doing something similar to what is done above where we would more or less stop Dask from doing its normal task-scheduling thing, hand-write a bunch of control flow, hope nothing breaks, run custom code, and then have Dask take back control when we're done.  

Both are totally doable, but would require us to build something like dask-xgboost, and raise general concerns around memory management, resilience, diagnostics, load balancing, spilling to disk, and so forth.  We lose a lot of Dask's management when we switch into this mode.

So maybe Dask should start thinking more about collective actions.  This is coming up often enough that it probably deserves more attention.  That's a large conversation though and probably requires dedicated time from someone somewhat deeply familiar with Dask scheduler internals.

**Comment by datametrician** [comment]
I agree with this approach. Thrust has a good single GPU sorting, and I think UCX should help tremendously.  Any thoughts on book keeping, or was that more of a placeholder for if we need it?

**Comment by mrocklin** [comment]
There are two forms of book keeping that are relevant here:

1.  The slicing and concatenation of cudf dataframes.  This is largely a test of memory management.
2.  The administrative tracking of tasks within the scheduler.  This becomes more of an issue as we scale out to more nodes.

**Comment by mrocklin** [comment]
> There are currently two implementations for this, one in dask.dataframe and one in dask-cudf which uses a batcher sorting net. While most dask-cudf code has been removed in favor of the dask.dataframe implementations this sorting code has remained, mostly because I don't understand it fully, and don't know if there was a reason for this particular implementation.

> Why was this implementation chosen? Was this discussed somewhere? Alternatively @sklam, do you have any information here?

@sklam can you expand on the motivation behind your use of batcher sort net?  Why was this decided on rather than the approach take in the mainline dask dataframe codebase?

**Comment by sklam** [comment]
@mrocklin, I was trying to avoid the following of the mainline sort/shuffle:
- Storing the data onto the disk
- Output is imbalanced.  All equal keys in the same partition.  If the key distribution is heavily skewed, GPU memory may run out.

I chose the sorting network because:

- It's very easy to implement.
- The sorting network has a fixed structure given the number of partitions.  
- There are more parallel regions. (this one is according to my bad memory)
- The memory usage is predictable.  The network can be seen as multiple-stages.  Each stage has data-dependency on the previous stage only.  Given 1GB of data, the algorithm will roughly need 2GB (1GB for the input and 1GB for the output at each stage).
- It will rebalance the partitions.  It may produce fewer partitions and the partition size will not be more than the `max(input_partition_sizes)`

**Comment by mrocklin** [comment]
OK, so these concerns seem similar to the concerns we had while designing the various shuffle algorithms for dask dataframe/bag.  I'm not seeing anything here that is highly specific to GPU computation.  (please correct me if I'm wrong).

My inclination then is to try to unify both CPU and GPU around a single implemenatation, probably starting with the one in dask.dataframe, but then maybe we would consider batcher-sorting-networks as an alternative for both, rather than just for dask-cudf, after doing some more extensive benchmarking.

Thanks for the information @sklam , I really appreciate the continuity here.

**Comment by mrocklin** [comment]
I recently took a look at Dask dataframe's task-based shuffle and improved docstrings here in order to help others dive in: https://github.com/dask/dask/pull/4674/files

I think that we need the following:

-  Something like the `pandas.util.hash_pandas_object` function, which hashes a pandas object row-by-row returning a Series of integer values.  (@kkraus14 do we have this?)

-  A `Series.searchsorted` method, allowing us to figure out where each row should go.  Used here:

    ```python
    def set_partitions_pre(s, divisions):
        partitions = pd.Series(divisions).searchsorted(s, side='right') - 1
        partitions[(s >= divisions[-1]).values] = len(divisions) - 2
        return partitions
    ```

    This helps us to assign a partition number to every row, based on where the future-index value of that row sits relative to the divisions.  Example: "Is the value in between the first and second divisions?  Great, it goes in the first partition."

-  The `dask.dataframe.shuffle.shuffle_group` function, which splits a pandas dataframe apart into a dict of pandas dataframes based on the value of a particular column provided by the hash values above.  We also need to modify this value in a particular way to achieved multi-staged shuffling.  This logic is explained in the PR above.

    We can either rewrite this from scratch or, if cudf is supporting ufuncs we might be able to get away with just dispatching on the `pandas._libs.algos.groupsort_indexer` and reuse all of the fancy logic here (which would be nice if cudf supports most ufuncs and out parameters)
-  Eventually we will also need quantile information.  This is also evolving in Dask core at the moment, so I suggest that we wait on this for a bit.

**Comment by mrocklin** [comment]
Also, just to direct people, I think that the core function that we'll have to make work is `rearrange_by_column_tasks`.  I think that all of set_index/merge/sort_values can be made to rely on functionality there.

**Comment by mt-jones** [comment]
@mrocklin @datametrician I have a vested interest in seeing this work succeed, and will begin by implementing the solution path you've outlined above. If we _need_ to get into the specifics of MPI/NCCL, we can cross that bridge later. I think a general solution which puts Dask first is going to help the ecosystem the most.

**Comment by mrocklin** [comment]
A minimal test would look something like the following:

```python
import cudf, dask.dataframe as dd

# Make a dataframe, we'd like to divide the data by y
df = cudf.DataFrame({'x': [1, 2, 3, 4, 5, 6], 'y': [0, 4, 0, 4, 0, 4]})

# Split it up into a few partitions with Dask
ddf = dd.from_pandas(df, npartitions=3)

# Try to create a new dataframe with two partitions sorted on the y column, split by y 0->2 and 2->4
dd.shuffle.rearrange_by_divisions(ddf, column='y', divisions=[0, 2, 4], shuffle='tasks')
```

**Comment by mrocklin** [comment]
I just tried this and ran into an minor problem of cudf.DataFrame.drop not supporting the axis= keyword.  As a suggestion, these errors are easier to identify if you remove some of Dask's error reporting with the following diff

```diff
diff --git a/dask/dataframe/core.py b/dask/dataframe/core.py
index 6a08af9..894fba6 100644
--- a/dask/dataframe/core.py
+++ b/dask/dataframe/core.py
@@ -3736,8 +3736,8 @@ def _emulate(func, *args, **kwargs):
     Apply a function using args / kwargs. If arguments contain dd.DataFrame /
     dd.Series, using internal cache (``_meta``) for calculation
     """
-    with raise_on_meta_error(funcname(func), udf=kwargs.pop('udf', False)):
-        return func(*_extract_meta(args, True), **_extract_meta(kwargs, True))
+    kwargs.pop('udf')
+    return func(*_extract_meta(args, True), **_extract_meta(kwargs, True))
```

**Comment by mrocklin** [comment]
https://github.com/rapidsai/cudf/issues/1074

**Comment by mrocklin** [comment]
I imagine that, like with the groupby aggregations work, this will end up triggering many small PRs in cudf.

**Comment by mrocklin** [comment]
Potential fix here: https://github.com/rapidsai/cudf/pull/1396

**Comment by mrocklin** [comment]
Next thing I run into, searchsorted

```python
In [1]: import cudf, dask.dataframe as dd
   ...:
   ...: # Make a dataframe, we'd like to divide the data by y
   ...: df = cudf.DataFrame({'x': [1, 2, 3, 4, 5, 6], 'y': [0, 4, 0, 4, 0, 4]})
   ...:
   ...: # Split it up into a few partitions with Dask
   ...: ddf = dd.from_pandas(df, npartitions=3)
   ...:
   ...: # Try to create a new dataframe with two partitions sorted on the y column, split by y 0->2 and 2->4
   ...: dd.shuffle.rearrange_by_divisions(ddf, column='y', divisions=[0, 2, 4], shuffle='tasks')
Out[1]: <dask_cudf.DataFrame | 40 tasks | 2 npartitions>

In [2]: _.compute(scheduler='single-threaded')
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
... <removed for clarity>
TypeError: can't compute boolean for <class 'cudf.dataframe.series.Series'>

In [3]: debug
> /home/nfs/mrocklin/cudf/python/cudf/dataframe/series.py(325)__bool__()
    323         into a boolean.
    324         """
--> 325         raise TypeError("can't compute boolean for {!r}".format(type(self)))
    326
    327     def values_to_string(self, nrows=None):

ipdb> up
> /home/nfs/mrocklin/miniconda/envs/cudf/lib/python3.7/site-packages/pandas/core/series.py(2337)searchsorted()
   2335             sorter = ensure_platform_int(sorter)
   2336         result = self._values.searchsorted(Series(value)._values,
-> 2337                                            side=side, sorter=sorter)
   2338
   2339         return result[0] if is_scalar(value) else result

ipdb>
> /home/nfs/mrocklin/dask/dask/dataframe/shuffle.py(434)set_partitions_pre()
    432
    433 def set_partitions_pre(s, divisions):
--> 434     partitions = pd.Series(divisions).searchsorted(s, side='right') - 1
    435     partitions[(s >= divisions[-1]).values] = len(divisions) - 2
    436     return partitions

ipdb>
```
